Implementation of `Identity` serivce

uselessgoddess opened the pull request.

uselessgoddess committed:
Initial vision of `Identity` service

uselessgoddess committed:
Add secret jwt key

uselessgoddess committed:
Add `MediatR` and `FluentValidation`:
- add `Walidate` attribute to enable/disable Validation
- add `Router` helper to shrink future Controllers size
- add way to enable commands validation by single command
- remove useless models from `Models` because they are in `Commands` now

uselessgoddess committed:
Add users' roles manipulation:
- add Users controllers:
  - `promote`, `demote` roles
  - add simple roles
  - add prepared `admin` user
- add new way to use `.Route` with exceptions instead of `null`:
  - add custom validation exceptions

uselessgoddess commented:
Add the following later: 
- email confirmation 
- password reset
- 2fa auth
- access policies 
- logging and caching

uselessgoddess committed:
Fix auth and roles validation+claims:
- fix useless field at `RefreshHandler`
- add `revoke` endpoint to revoke refresh-tokens
- change layout of JWT token
- add `Admin` policy
- add roles validation and change type from uint to string

uselessgoddess committed:
Fix some mismatching:
- use `Username` everywhere
- remove unused imports
- add `/user`, `/users` routes
- fix controllers https commands

uselessgoddess committed:
Add StyleCop and split files by CQRS arch:
- split models to files
- move users/user to `Queries`

uselessgoddess committed:
Use Middlewares to handle exception and validation:
- move out validation and exception catching from `Route` into Middlewares
- `Users.Count` -> `Users.Size` in famous of symmetry
- fix names of controller's methods

uselessgoddess committed:
Fix password validation options

uselessgoddess committed:
Fix Exceptions namespace

MefistoSamael commented on Identity/Core/Models/RefreshToken.cs:
move to domain layer

MefistoSamael replied:
place empty lines between methods

MefistoSamael commented on Identity/Core/Models/AppUser.cs:
move to domain

MefistoSamael commented on Identity/Core/Models/Role.cs:
move to domain

MefistoSamael commented on Identity/Core/Commands/Login.cs:
```diff
@@ -0,0 +1,23 @@
+
+public class LoginValidation : AbstractValidator<Login>
+{
+    public LoginValidation()
+    {
+        RuleFor(x => x.Username)
+            .NotEmpty();
+
+        RuleFor(x => x.Password)
+            .NotEmpty();
+    }
```
separate all validators and DTO's

uselessgoddess replied:
but this isn't DTO))

MefistoSamael commented on Identity/Core/Commands/Validation.cs:
rename class. Move in root of core

MefistoSamael commented on Identity/Core/Handlers/DemoteHandler.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    public async Task Handle(Demote req, CancellationToken cancellationToken)
+    {
+        var user = await users.FindByIdAsync(req.Id) ?? throw new NotFoundException();
+        await users.RemoveFromRolesAsync(user, req.Roles.Select(r => r.ToString()));
```
await must be preceeded and followed by blank lines

uselessgoddess replied:
```suggestion
        var user = await users.FindByIdAsync(req.Id) ?? throw new NotFoundException();

        await users.RemoveFromRolesAsync(user, req.Roles.Select(r => r.ToString()));
```

MefistoSamael commented on Identity/Core/Handlers/LoginHandler.cs:
```diff
@@ -0,0 +1,36 @@
+
+        if (!result.Succeeded)
+        {
+            throw new UnauthorizedException("invalid username or password");
```
it is badrequest

uselessgoddess replied:
```suggestion
            throw new BadRequestException("invalid username or password");
```

MefistoSamael commented on Identity/Core/Handlers/RefreshHandler.cs:
```diff
@@ -0,0 +1,43 @@
+
+        if (old == null || old.Expire < DateTime.Now || old.IsRevoked)
+        {
+            throw new UnauthorizedException();
```
use badrequest exception

uselessgoddess replied:
```suggestion
            throw new BadRequestException();
```

MefistoSamael commented on Identity/Core/Handlers/RegisterHandler.cs:
```diff
@@ -0,0 +1,27 @@
+        if (!(await users.CreateAsync(user, req.Password)).Succeeded)
+        {
+            throw new BadRequestException("user already exists");
+        }
```
1. conditional statements must be followed and preceeded by blank lines

2. Use alreadyExists exception

MefistoSamael commented on Identity/Core/Handlers/RegisterHandler.cs:
```diff
@@ -0,0 +1,27 @@
+    {
+        if (await users.FindByEmailAsync(req.Email) == null)
+        {
+            throw new ValidationException(new[]
+                { new ValidationFailure("Email", "user with the same email already exists") });
```
use badRequest exception

uselessgoddess replied:
```suggestion
            throw new BadRequestException("user with the same email already exists");
```

MefistoSamael commented on Identity/Core/QueryHandlers/UsersHandlers.cs:
1. separate handlers into different files
2. use more descriptive names for handlers

uselessgoddess replied:
1. ok
2. names are subjective, but their len saves the memory

MefistoSamael replied:
you shoudn't care so much about memory in such types of application

MefistoSamael replied:
at first you should care about maintainability and extendibility

MefistoSamael commented on Identity/Infrastructure/Data/DatabaseCx.cs:
use more descriptive name

uselessgoddess replied:
how can I improve descriptive of `Database Context`

MefistoSamael commented on Identity/Infrastructure/Utils.cs:
seed db with HasData method

uselessgoddess replied:
but why? my way is easier and hides low-level representation

MefistoSamael replied:
naming conventions) ¯\_(ツ)_/¯

MefistoSamael commented on Identity/Program.cs:
separate services configuration in different files by layers

MefistoSamael reviewed (changes requested):
Each layer of application has to be created as class library(except presentation)

uselessgoddess committed:
Apply simple suggestions from code review

uselessgoddess committed:
Review bonks:
- fix some style typos
- split some classes
- use `.HasData()` seeds:
  - add admin role migration
- use di styled app building

uselessgoddess committed:
`Core`, `Infrastructure` as class libraries

uselessgoddess committed:
Completely split layers into different projects

uselessgoddess committed:
More descriptive names, add missing `launchSettings.json`

uselessgoddess committed:
Add forgotten admin seeds

uselessgoddess committed:
Split exceptions, fix AlreadyExistsException, remove useless middleware

uselessgoddess committed:
Fix and reorder imports

MefistoSamael commented on Identity/Api/Program.cs:
```diff
@@ -0,0 +1,74 @@
+builder.Services.AddControllers();
+builder.Services.AddDataValidation();
+
+builder.Services.AddAuthorizationBuilder()
+    .AddPolicy(Policy.Admin, policy => policy.RequireRole("Admin"));
+
+builder.Services.AddSwaggerGen(c =>
+{
+    c.SwaggerDoc("v1", new OpenApiInfo { Title = "IdentityService API", Version = "v1" });
+
+    c.AddSecurityDefinition("Bearer", new OpenApiSecurityScheme
+    {
+        Name = "Authorization",
+        Type = SecuritySchemeType.ApiKey,
+        Scheme = "Bearer",
+        BearerFormat = "JWT",
+        In = ParameterLocation.Header,
+        Description = "Put JWT Token in Bearer format {token}",
+    });
+    c.AddSecurityRequirement(new OpenApiSecurityRequirement
+    {
+        {
+            new OpenApiSecurityScheme
+            {
+                Reference = new OpenApiReference
+                {
+                    Type = ReferenceType.SecurityScheme,
+                    Id = "Bearer",
+                },
+            },
+            []
+        },
+    });
+});
```
take it away from program

MefistoSamael commented on Identity/Api/App/Controllers/AccountController.cs:
separate business logic from presentation logic

uselessgoddess committed:
Migration into Clean architecture

uselessgoddess committed:
Merge remote-tracking branch 'origin/feat/identity-service' into feat/identity-service

# Conflicts:
#	Identity/Api/App/Handlers/RevokeHandler.cs
#	Identity/Api/Middlewares/Middlewares.cs
#	Identity/Api/Middlewares/ValidationMiddleware.cs
#	Identity/UseCases/Commands/Handlers/LoginHandler.cs
#	Identity/UseCases/Commands/Handlers/RefreshHandler.cs

uselessgoddess committed:
Post-merge mismatches:
- fix missing migration file
- fix some imports
- fix exceptions constructors

uselessgoddess committed:
Push missing designer migration's file

MefistoSamael commented on Identity/Api/Program.cs:
```diff
@@ -0,0 +1,46 @@
+builder.Services.AddControllers();
+builder.Services.AddDataValidation();
+
+builder.Services.AddSwaggerGen(c => { c.AddSecurityDefinition("Bearer", new OpenApiSecurityScheme()); });
```
move to api configuration

MefistoSamael commented on Identity/Api/DependencyInjection.cs:
```diff
@@ -0,0 +1,38 @@
+        builder.Services.AddAuthorizationBuilder()
+            .AddPolicy(Policy.Admin, policy => policy.RequireRole("Admin"));
+
+        return builder;
```
move to infrastructure configuration

MefistoSamael commented on Identity/UseCases/Services/TokenService.cs:
move to infrastructure

uselessgoddess committed:
Fix dependency injection from review requests
